### PR TITLE
Bugfix in temporal buffer

### PR DESCRIPTION
**Pull request by mfehr**
There was a small bug in `extractItemsBeforeIncluding`, it was extracting  value that was newer than the threshold. The consequence was that the synchronizer of the maplab_node was releasing data a bit early sometimes. The corner cases should now be covered by the unit tests.

### CODE REVIEW

File: common/maplab-common/include/maplab-common/temporal-buffer-inl.h [hunk 1]
```diff
@@ -391,69 +391,63 @@ size_t TemporalBuffer<ValueType, AllocatorType>::extractItemsBeforeIncluding(
...
 
-    // If all values in the buffer should be removed we don't need to move the
-    // iterator another step.
-    CHECK(it != values_.begin());
-    if (it != values_.end()) {
-      it = std::next(it);
-    }
+  // Get first iterator that is greater than the timestamp, i.e. everything
+  // between begin() and this iterator is what we want.
+  auto first_newer_timestamp_it = values_.upper_bound(timestamp_ns);
```
Comment (schneith) [inline common/maplab-common/include/maplab-common/temporal-buffer-inl.h:403]:
```suggestion
  typename BufferType::iterator first_newer_timestamp_it =
		  values_.upper_bound(timestamp_ns);
```

File: common/maplab-common/include/maplab-common/temporal-buffer-inl.h [hunk 2]
```diff
@@ -391,69 +391,63 @@ size_t TemporalBuffer<ValueType, AllocatorType>::extractItemsBeforeIncluding(
...
 
-    const size_t size_before = values_.size();
+  // Copy values to be removed.
+  std::transform(
+      values_.begin(), first_newer_timestamp_it,
+      std::back_inserter(*removed_values),
+      [](auto stamped_value) { return stamped_value.second; });
```
Comment (schneith) [inline common/maplab-common/include/maplab-common/temporal-buffer-inl.h:409]:
```suggestion
  std::transform(
      values_.begin(), first_newer_timestamp_it,
      std::back_inserter(*removed_values),
      [](const typename BufferType::value_type& stamped_value) {
	  return stamped_value.second;
  });
```

File: common/maplab-common/include/maplab-common/temporal-buffer-inl.h [hunk 3]
```diff
@@ -391,69 +391,63 @@ size_t TemporalBuffer<ValueType, AllocatorType>::extractItemsBeforeIncluding(
...
+  // return.
+  if (timestamp_ns < values_.begin()->first) {
+    return 0u;
   }
-  return 0u;
+  // Otherwise there is at least one element that is equal or older.
+
+  // Get first iterator that is greater than the timestamp, i.e. everything
+  // between begin() and this iterator is what we want.
+  auto first_newer_timestamp_it = values_.upper_bound(timestamp_ns);
```
Comment (schneith) [inline common/maplab-common/include/maplab-common/temporal-buffer-inl.h:437]:
```suggestion
  typename BufferType::iterator first_newer_timestamp_it =
		  values_.upper_bound(timestamp_ns);
```

File: common/maplab-common/include/maplab-common/temporal-buffer-inl.h [hunk 4]
```diff
@@ -391,69 +391,63 @@ size_t TemporalBuffer<ValueType, AllocatorType>::extractItemsBeforeIncluding(
...
+
+  // Copy values earlier or equal to the timestamp.
+  std::transform(
+      values_.begin(), first_newer_timestamp_it,
+      std::back_inserter(*returned_values),
+      [](auto stamped_value) { return stamped_value.second; });
```
Comment (schneith) [inline common/maplab-common/include/maplab-common/temporal-buffer-inl.h:443]:
```suggestion
  std::transform(
      values_.begin(), first_newer_timestamp_it,
      std::back_inserter(*returned_values),
      [](const typename BufferType::value_type& stamped_value) {
	  return stamped_value.second;
  });
```
